slsDetectorSoftware/src/Module.cpp

```diff
@@ -1036,24 +1027,64 @@ void Module::setNumberOfDigitalSamples(int value) {
     }
 }
 
-int64_t Module::getExptime() { return sendToDetector<int64_t>(F_GET_EXPTIME); }
+int Module::getNumberOfGates() { return sendToDetector<int>(F_GET_NUM_GATES); }
+
+void Module::setNumberOfGates(int value) {
+    LOG(logDEBUG1) << "Setting number of gates to " << value;
+    sendToDetector(F_SET_NUM_GATES, value, nullptr);
+    if (shm()->useReceiverFlag) {
+        LOG(logDEBUG1) << "Sending number of gates to Receiver: " << value;
+        sendToReceiver(F_SET_RECEIVER_NUM_GATES, value, nullptr);
+    }
+}
+
+int64_t Module::getExptime(int gateIndex) {
+    return sendToDetector<int64_t>(F_GET_EXPTIME, gateIndex);
+}
 
-void Module::setExptime(int64_t value) {
+void Module::setExptime(int gateIndex, int64_t value) {
     int64_t prevVal = value;
     if (shm()->myDetectorType == EIGER) {
-        prevVal = getExptime();
+        prevVal = getExptime(-1);
     }
-    LOG(logDEBUG1) << "Setting exptime to " << value << "ns";
-    sendToDetector(F_SET_EXPTIME, value, nullptr);
+    LOG(logDEBUG1) << "Setting exptime to " << value
+                   << "ns (gateindex: " << gateIndex << ")";
+    int64_t args[]{static_cast<int64_t>(gateIndex), value};
+    sendToDetector(F_SET_EXPTIME, args, nullptr);
     if (shm()->useReceiverFlag) {
         LOG(logDEBUG1) << "Sending exptime to Receiver: " << value;
-        sendToReceiver(F_RECEIVER_SET_EXPTIME, value, nullptr);
+        sendToReceiver(F_RECEIVER_SET_EXPTIME, args, nullptr);
     }
     if (prevVal != value) {
         updateRateCorrection();
```

Comment:
Not for this PR but an idea, could this check be done in the server? Meaning that every time we set exptime the ratecorr gets updated. (if enabled) and the server tracks to see if it's a new exptime or not. Then we need less logic in the client

Reply:
This could, but I wanted it to be also changed in the receiver, before doing updateratecorrection. If there is an error in ratecorrection, which almost certainly could be (if dr is not 16 or 32), then the receiver does not get updated because of the exception. This is why I put it here in the client, after all the exptime is actually set for detector and receiver.